<h1>Activity: PENGUINS</h1>

<h2>Sample</h2>

In [1]:
#libraries

In [2]:
import numpy as np

In [3]:
import pandas as pd

In [4]:
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)


In [5]:
bird = pd.read_csv('datasets\\penguins.csv')

In [6]:
bird

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
1       Adelie  Torgersen            39.5           17.4              186.0   
2       Adelie  Torgersen            40.3           18.0              195.0   
3       Adelie  Torgersen             NaN            NaN                NaN   
4       Adelie  Torgersen            36.7           19.3              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[344 rows x 8 columns]

<h1>DESCRIPTIVE ANALYTICS</h1>

In [7]:
bird.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
count      342.000000     342.000000         342.000000   342.000000   
mean        43.921930      17.151170         200.915205  4201.754386   
std          5.459584       1.974793          14.061714   801.954536   
min         32.100000      13.100000         172.000000  2700.000000   
25%         39.225000      15.600000         190.000000  3550.000000   
50%         44.450000      17.300000         197.000000  4050.000000   
75%         48.500000      18.700000         213.000000  4750.000000   
max         59.600000      21.500000         231.000000  6300.000000   

              year  
count   344.000000  
mean   2008.029070  
std       0.818356  
min    2007.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2009.000000

In [8]:
bird.describe().T

count         mean         std     min       25%      50%  \
bill_length_mm     342.0    43.921930    5.459584    32.1    39.225    44.45   
bill_depth_mm      342.0    17.151170    1.974793    13.1    15.600    17.30   
flipper_length_mm  342.0   200.915205   14.061714   172.0   190.000   197.00   
body_mass_g        342.0  4201.754386  801.954536  2700.0  3550.000  4050.00   
year               344.0  2008.029070    0.818356  2007.0  2007.000  2008.00   

                      75%     max  
bill_length_mm       48.5    59.6  
bill_depth_mm        18.7    21.5  
flipper_length_mm   213.0   231.0  
body_mass_g        4750.0  6300.0  
year               2009.0  2009.0

In [9]:
bird["species"]

0         Adelie
1         Adelie
2         Adelie
3         Adelie
4         Adelie
         ...    
339    Chinstrap
340    Chinstrap
341    Chinstrap
342    Chinstrap
343    Chinstrap
Name: species, Length: 344, dtype: object

In [10]:
bird["species"] == "Adelie"

0       True
1       True
2       True
3       True
4       True
       ...  
339    False
340    False
341    False
342    False
343    False
Name: species, Length: 344, dtype: bool

In [11]:
bird[bird["species"] == "Adelie"]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
147  Adelie      Dream            36.6           18.4              184.0   
148  Adelie      Dream            36.0           17.8              195.0   
149  Adelie      Dream            37.8           18.1              193.0   
150  Adelie      Dream            36.0           17.1              187.0   
151  Adelie      Dream            41.5           18.5              201.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
147       3475.0  female  2009  
148       3450.0  female  2009  
149       3750.0    male  2009  
150       3700.0  female  2009  
151       4000.0    male  2009  

[152 rows x 8 columns]

In [12]:
adelie_df = bird[bird["species"] == "Adelie"]

In [13]:
adelie_df

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
147  Adelie      Dream            36.6           18.4              184.0   
148  Adelie      Dream            36.0           17.8              195.0   
149  Adelie      Dream            37.8           18.1              193.0   
150  Adelie      Dream            36.0           17.1              187.0   
151  Adelie      Dream            41.5           18.5              201.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
147       3475.0  female  2009  
148       3450.0  female  2009  
149       3750.0    male  2009  
150       3700.0  female  2009  
151       4000.0    male  2009  

[152 rows x 8 columns]

In [14]:
bird['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

In [15]:
gentoo_df = bird[bird["species"] == "Gentoo"]

In [16]:
chinstrap_df = bird[bird["species"] == "Chinstrap"]

In [17]:
gentoo_df

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
152  Gentoo  Biscoe            46.1           13.2              211.0   
153  Gentoo  Biscoe            50.0           16.3              230.0   
154  Gentoo  Biscoe            48.7           14.1              210.0   
155  Gentoo  Biscoe            50.0           15.2              218.0   
156  Gentoo  Biscoe            47.6           14.5              215.0   
..      ...     ...             ...            ...                ...   
271  Gentoo  Biscoe             NaN            NaN                NaN   
272  Gentoo  Biscoe            46.8           14.3              215.0   
273  Gentoo  Biscoe            50.4           15.7              222.0   
274  Gentoo  Biscoe            45.2           14.8              212.0   
275  Gentoo  Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  year  
152       4500.0  female  2007  
153       5700.0    male  2007  
154       4450.0  female  2007  
155       5700.0    male  2007  
156       5400.0    male  2007  
..           ...     ...   ...  
271          NaN     NaN  2009  
272       4850.0  female  2009  
273       5750.0    male  2009  
274       5200.0  female  2009  
275       5400.0    male  2009  

[124 rows x 8 columns]

In [18]:
chinstrap_df

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
276  Chinstrap  Dream            46.5           17.9              192.0   
277  Chinstrap  Dream            50.0           19.5              196.0   
278  Chinstrap  Dream            51.3           19.2              193.0   
279  Chinstrap  Dream            45.4           18.7              188.0   
280  Chinstrap  Dream            52.7           19.8              197.0   
..         ...    ...             ...            ...                ...   
339  Chinstrap  Dream            55.8           19.8              207.0   
340  Chinstrap  Dream            43.5           18.1              202.0   
341  Chinstrap  Dream            49.6           18.2              193.0   
342  Chinstrap  Dream            50.8           19.0              210.0   
343  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
276       3500.0  female  2007  
277       3900.0    male  2007  
278       3650.0    male  2007  
279       3525.0  female  2007  
280       3725.0    male  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[68 rows x 8 columns]

In [19]:
adelie_df["body_mass_g"]

0      3750.0
1      3800.0
2      3250.0
3         NaN
4      3450.0
        ...  
147    3475.0
148    3450.0
149    3750.0
150    3700.0
151    4000.0
Name: body_mass_g, Length: 152, dtype: float64

In [20]:
gentoo_df["body_mass_g"]

152    4500.0
153    5700.0
154    4450.0
155    5700.0
156    5400.0
        ...  
271       NaN
272    4850.0
273    5750.0
274    5200.0
275    5400.0
Name: body_mass_g, Length: 124, dtype: float64

In [21]:
chinstrap_df["body_mass_g"]

276    3500.0
277    3900.0
278    3650.0
279    3525.0
280    3725.0
        ...  
339    4000.0
340    3400.0
341    3775.0
342    4100.0
343    3775.0
Name: body_mass_g, Length: 68, dtype: float64

In [22]:
adelie_df["body_mass_g"].mean()

np.float64(3700.662251655629)

In [23]:
gentoo_df["body_mass_g"].mean()

np.float64(5076.016260162602)

In [24]:
chinstrap_df["body_mass_g"].mean()

np.float64(3733.0882352941176)

In [25]:
bird[(bird["species"] == "Adelie") & (bird["sex"] == "female")]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
4    Adelie  Torgersen            36.7           19.3              193.0   
6    Adelie  Torgersen            38.9           17.8              181.0   
12   Adelie  Torgersen            41.1           17.6              182.0   
..      ...        ...             ...            ...                ...   
142  Adelie      Dream            32.1           15.5              188.0   
144  Adelie      Dream            37.3           16.8              192.0   
147  Adelie      Dream            36.6           18.4              184.0   
148  Adelie      Dream            36.0           17.8              195.0   
150  Adelie      Dream            36.0           17.1              187.0   

     body_mass_g     sex  year  
1         3800.0  female  2007  
2         3250.0  female  2007  
4         3450.0  female  2007  
6         3625.0  female  2007  
12        3200.0  female  2007  
..           ...     ...   ...  
142       3050.0  female  2009  
144       3000.0  female  2009  
147       3475.0  female  2009  
148       3450.0  female  2009  
150       3700.0  female  2009  

[73 rows x 8 columns]

In [26]:
bird.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [27]:
bird['bill_length_mm'] = bird['bill_length_mm'].fillna(bird['bill_length_mm'].mean())
bird['bill_depth_mm'] = bird['bill_depth_mm'].fillna(bird['bill_depth_mm'].mean())
bird['flipper_length_mm'] = bird['flipper_length_mm'].fillna(bird['flipper_length_mm'].mean())
bird['body_mass_g'] = bird['body_mass_g'].fillna(bird['body_mass_g'].mean())
bird['sex'] = bird['sex'].fillna(bird['sex'].mode()[0])

In [28]:
bird.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

In [29]:
bird['species']= bird['species'].astype('category')
bird['island']= bird['island'].astype('category')
bird['sex']= bird['sex'].astype('category')

In [30]:
bird.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    category
 2   bill_length_mm     344 non-null    float64 
 3   bill_depth_mm      344 non-null    float64 
 4   flipper_length_mm  344 non-null    float64 
 5   body_mass_g        344 non-null    float64 
 6   sex                344 non-null    category
 7   year               344 non-null    int64   
dtypes: category(3), float64(4), int64(1)
memory usage: 15.0 KB


<h3>1. What is the highest body mass of Female Adelie species in Dream Island?</h3>

In [31]:
adelie_female = bird[(bird["species"] == "Adelie") & (bird["sex"] == "female")]
adelie_dream = adelie_female[adelie_female["island"] == "Dream"]
adelie_dream= adelie_dream["body_mass_g"].max()
adelie_dream

np.float64(3700.0)

<h4>According to the given data, the highest body mass of Female Adelie species in Dream Island is 3700</h4>

<h3>2. Are there any penguins with a bill depth greater than 21mm?</h3>

In [32]:
bird_max_depth = bird[bird["bill_depth_mm"] > 21]
bird_max_depth

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
13  Adelie  Torgersen            38.6           21.2              191.0   
14  Adelie  Torgersen            34.6           21.1              198.0   
19  Adelie  Torgersen            46.0           21.5              194.0   
35  Adelie      Dream            39.2           21.1              196.0   
49  Adelie      Dream            42.3           21.2              191.0   
61  Adelie     Biscoe            41.3           21.1              195.0   

    body_mass_g   sex  year  
13       3800.0  male  2007  
14       4400.0  male  2007  
19       4200.0  male  2007  
35       4150.0  male  2007  
49       4150.0  male  2007  
61       4400.0  male  2008

<h4>According to the given data, there is a bill depth greater than 21mm</h4>

<h3>3. What is the highest recorded body mass in all male species?</h3>

In [33]:
male_df = bird[bird["sex"] == "male"]
max_mass_male_df = male_df["body_mass_g"].max()
max_mass_male_df

np.float64(6300.0)

<h4>According to the given data, the highest recorded bodymass in all male species is 6300</h4>

<h3>4. What is the highest recorded bill length among the female species?</h3>

In [34]:
female_df = bird[bird["sex"] == "female"]
max_bill_female = female_df["bill_length_mm"].max()
max_bill_female

np.float64(58.0)

<h4>According to the given data, the highest recorded bill length among the female species is 58</h4>

<h3>5. In all of the penguins in Biscoe Island, what penguin breed mostly has high body mass?</h3>

In [35]:
biscoe_adelie = bird[(bird["island"] == "Biscoe") & (bird["species"] == "Adelie")]
biscoe_adelie_mass= biscoe_adelie["body_mass_g"].mean()
biscoe_adelie_mass

np.float64(3709.659090909091)

In [36]:
biscoe_chinstrap = bird[(bird["island"] == "Dream") & (bird["species"] == "Chinstrap")]
biscoe_chinstrap_mass= biscoe_chinstrap["body_mass_g"].mean()
biscoe_chinstrap_mass

np.float64(3733.0882352941176)

In [37]:
biscoe_gentoo = bird[(bird["island"] == "Biscoe") & (bird["species"] == "Gentoo")]
biscoe_gentoo_mass= biscoe_gentoo["body_mass_g"].mean()
biscoe_gentoo_mass

np.float64(5068.965761177136)

<h4>According to the given data, the penguin breed with the most high body mass is Gentoo Species</h4>

<h3>6. Is there any male penguins in Biscoe Island which have a 3000g or less body mass</h3>

In [38]:
biscoe_m = bird[(bird["island"] == "Biscoe") & (bird["sex"] == "male")]
biscoe_m_mass= biscoe_m["body_mass_g"].min()
biscoe_m_mass

np.float64(3550.0)

<h4>Based on the given data. The minimum body mass in Biscoe Island is 3550g, concluding that there is no Male in Biscoe Island which has 3000g body mass </h4>

<h3>7. Do you think there's an species of penguin that have 50mm or higher Bill Length and 230mm or higher Flipper Length?</h3
                                                                                                                            >

In [39]:
bird_bill = bird[(bird["bill_length_mm"] > 50) & (bird["flipper_length_mm"] > 230)]
bird_bill.value_counts()

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  sex   year
Gentoo   Biscoe  54.3            15.7           231.0              5650.0       male  2008    1
Name: count, dtype: int64

<h4>According to the data, there is a Penguin having 50mm or higher Bill Length and 230mm or higher Flipper Length </h4>

<h3>8. What is the least dominant species in the Dream Island?</h3>

In [40]:
bird_dream = bird[bird["island"] == "Dream"]

species_count = bird_dream["species"].value_counts()
species_count

species
Chinstrap    68
Adelie       56
Gentoo        0
Name: count, dtype: int64

<h4>Based on the data, the least dominant species in the Dream Island is Gentoo species having 0 record</h4>

<h3>9. Do you think there is any fluctuation or changes of body mass of the penguins on those 3 years?</h3>

In [41]:
bird_body_mass = bird.groupby("year")["body_mass_g"].mean()
bird_body_mass

year
2007    4125.243222
2008    4266.666667
2009    4210.222953
Name: body_mass_g, dtype: float64

<h4>According to the data, there is slight fluctuation of body weight of the penguins within those three years. In 2007, the recorded average body mass of penguins was 4125 but it increase was noticeable in the year 2008 reaching 4266, and decreased slightly in 2009 having 4210.</h4>

<h3>10. In terms of flipper length, does it mostly depends on the species or the sex?</h3>

In [44]:
bird_species_flipper = bird.groupby("species")["flipper_length_mm"].mean()

bird_species_sex = bird.groupby("sex")["flipper_length_mm"].mean()

bird_species_sex_flipper = bird.groupby(["species", "sex"])["flipper_length_mm"].mean()

bird_species_sex_flipper

species    sex   
Adelie     female    187.794521
           male      192.087534
Chinstrap  female    191.735294
           male      199.911765
Gentoo     female    212.706897
           male      220.877503
Name: flipper_length_mm, dtype: float64

<h4>Based on the data above, species is the main factor influencing flipper length. Sex has a smaller effect within each species as males have slightly longer flipper length than females. Gentoo males have the longest flipper length, and the species is a higher factor than the effect of sex when analyzing the flipper length.</h4>

<h3>11. Supporting the analysis on the question no.10, is it also possible that the island they live in is involve?</h3>

In [43]:
bird_island_bodymass_ave = bird.groupby("island")["body_mass_g"].mean()
bird_island_bodymass_ave

island
Biscoe       4712.956871
Dream        3712.903226
Torgersen    3715.899123
Name: body_mass_g, dtype: float64

<h4>According to the data, it is highly possible that the island has an effect on the average body mass of a penguin. The data shows that Biscoe holds the highest penguin average body mass, while Dream and Torgersen are on a closer gap.</h4>

<h3>12. Is greater bill length greatly affects the body mass of a penguin?" </h3>

In [47]:
mean_bill_length = bird["bill_length_mm"].mean()

bird_greater_bill = bird[bird["bill_length_mm"] > mean_bill_length]["body_mass_g"].mean()
bird_greater_bill

np.float64(4634.714285714285)

In [46]:
bird_lesser_bill = bird[bird["bill_length_mm"] <= mean_bill_length]["body_mass_g"].mean()
bird_lesser_bill

np.float64(3753.4231288279875)

<h4>According to the data, the greater bill length can actually affect the body mass of a penguin</h4>